sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala

```diff
@@ -274,7 +275,21 @@ object InMemoryFileIndex extends Logging {
     // [SPARK-17599] Prevent InMemoryFileIndex from failing if path doesn't exist
     // Note that statuses only include FileStatus for the files and dirs directly under path,
     // and does not include anything else recursively.
-    val statuses = try fs.listStatus(path) catch {
+    val statuses: Array[FileStatus] = try {
+      fs match {
+        // DistributedFileSystem overrides listLocatedStatus to make 1 single call to namenode
+        // to retrieve the file status with the file block location. The reason to still fallback
+        // to listStatus is because the default implementation would potentially throw a
+        // FileNotFoundException which is better handled by doing the lookups manually below.
+        case _: DistributedFileSystem =>
```

Comment:
@melin Are you hitting a perf regression? Do we need a SQLConf? 

cc @JoshRosen

Reply:
Unless I'm misunderstanding something, I don't think that @melin's posted benchmark is an apples-to-apples comparison: `listLocatedStatus` is going to be slower than `listStatus` because it's doing more work in order to get location information. In order for this to be a fair performance comparison you'd want to compare `listStatus + getBlockLocations` to `listLocatedStatus`.

The direct comparison is only useful in cases where we're never going to make use of the location information and to my knowledge we currently do not have a flag to bypass locality fetching in `InMemoryFileIndex` (so the location-free performance isn't relevant).

Reply:
Yep. Of course, if there is a corner case, we had better have conf for this. For now, it looks like we need more description about what he is hitting. SPARK JIRA would be good for that.

Reply:
Before coming up with this change I was actually wondering why there was no conf options to ignore locality and why it was required to do those extra lookups. In situations where the spark job is running on a small set of servers out of a big HDFS cluster, locality is a bit useless as the hit rate is super low, the time to get block locations was getting out of control for us and locality wait per task was slowing down the jobs. 

I think this change to get the blocks from the namenode is good enough for us now but I can definitely see the benefit of having a conf option to disable locality entirely which would turn off fetching of getFileBlockLocations and also use listStatus always instead of my optimization to use listLocatedStatus.

Reply:
Please file a JIRA for the further discussions on new `conf` if it's not filed yet. (I didn't search it, but might be exist in some other context.)
> having a conf option to disable locality entirely which would turn off fetching of getFileBlockLocations